### ISSUE TITLE
osx-arm64 gcdump test fails (during rundown)

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/runtime#42022]
The gcdump.sh test is failing with an assert. (with #40435)

Observations:
The assert is triggered by the gc doing an SetMarked on a object without a method table.  
The GC was triggered by `EventPipeProtocolHelper::StopTracing`.  
The assert occurred during rundown.
osx-arm64 doesn't have an IPI implementation yet.

```
/Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/tracing/eventpipe/gcdump/gcdump/gcdump.sh

Assert failure(PID 49601 [0x0000c1c1], Thread: 690020 [0xa8764]): RawGetMethodTable()
    File: /Users/steve/git/runtime/src/coreclr/src/gc/gc.cpp Line: 3699
    Image: /Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/Tests/Core_Root/corerun

/Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/tracing/eventpipe/gcdump/gcdump/gcdump.sh: line 370: 49601 Abort trap: 6           $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
BEGIN EXECUTION
/Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/Tests/Core_Root/corerun gcdump.dll ''
  0.0s: ==TEST STARTING==
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47422-1599608288-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47312-1599600533-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47300-1599600522-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47349-1599600615-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47326-1599600539-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47353-1599600913-socket
  4.2s: Deleted
  4.2s: Attempting to delete the zombied pipe: /var/folders/9m/qdyqmjnd7ys_dx65n7yv4l180000gq/T/dotnet-diagnostic-47292-1599600520-socket
  4.2s: Deleted
  4.4s: Started sending sentinel events...
  4.5s: Connecting to EventPipe...
  6.5s: Connected to EventPipe with sessionID '0x132831780'
  6.5s: Creating EventPipeEventSource...
  6.9s: EventPipeEventSource created
  7.0s: Dynamic.All callback registered
  7.0s: Running optional trace validator
  7.7s: Finished running optional trace validator
  7.7s: Starting stream processing...
  8.2s: Saw new provider 'Microsoft-Windows-DotNETRuntime'
  8.7s: Saw sentinel event
  8.7s: Stopped sending sentinel events
  8.7s: Starting event generating action...
  8.7s: Stopping event generating action
  8.7s: Sending StopTracing command...
  9.0s: Saw new provider 'Microsoft-DotNETCore-EventPipe'
  9.0s: Saw new provider 'Microsoft-Windows-DotNETRuntimeRundown'
Expected: 100
Actual: 134
END EXECUTION - FAILED
```

```
 bt
* thread #4, stop reason = EXC_BREAKPOINT (code=1, subcode=0x1010b14e4)
    frame #0: 0x00000001010b14e4 libcoreclr.dylib`DBG_DebugBreak at debugbreak.S:7
    frame #1: 0x0000000101017d68 libcoreclr.dylib`::DebugBreak() at debug.cpp:410:9
    frame #2: 0x000000010168497c libcoreclr.dylib`::DbgAssertDialog(szFile="/Users/steve/git/runtime/src/coreclr/src/gc/gc.cpp", iLine=3699, szExpr="RawGetMethodTable()") at debug.cpp:697:13
    frame #3: 0x00000001016365c4 libcoreclr.dylib`WKS::CObjectHeader::SetMarked(this=0x000000010304e030) at gc.cpp:3699:9
    frame #4: 0x00000001016075cc libcoreclr.dylib`WKS::gc_heap::mark_object_simple1(unsigned char*, unsigned char*) [inlined] WKS::gc_heap::gc_mark1(o="") at gc.cpp:18457:5
    frame #5: 0x00000001016075a8 libcoreclr.dylib`WKS::gc_heap::mark_object_simple1(unsigned char*, unsigned char*) [inlined] WKS::gc_heap::gc_mark(o="", low="\x10&D\x03\x01", high="") at gc.cpp:18467
    frame #6: 0x000000010160757c libcoreclr.dylib`WKS::gc_heap::mark_object_simple1(oo="\x81\x02", start="\x81\x02") at gc.cpp:18894
    frame #7: 0x000000010160906c libcoreclr.dylib`WKS::gc_heap::mark_object_simple(po=0x000000016ff9d5a0) at gc.cpp:19437:17
    frame #8: 0x000000010160ddf4 libcoreclr.dylib`WKS::GCHeap::Promote(ppObject=0x00000001001b15f8, sc=0x000000016ff9dac8, flags=2) at gc.cpp:36359:5
    frame #9: 0x00000001015554d0 libcoreclr.dylib`PinObject(pObjRef=0x00000001001b15f8, pExtraInfo=0x0000000000000000, lp1=6173612744, lp2=4318092408) at objecthandle.cpp:268:5
    frame #10: 0x000000010155250c libcoreclr.dylib`ScanConsecutiveHandlesWithoutUserData(pValue=0x00000001001b15f8, pLast=0x00000001001b1600, pInfo=0x000000016ff9d858, (null)=0x0000000000000000) at handletablescan.cpp:443:13
    frame #11: 0x0000000101552710 libcoreclr.dylib`BlockScanBlocksWithoutUserData(pSegment=0x00000001001b0000, uBlock=2, uCount=1, pInfo=0x000000016ff9d858) at handletablescan.cpp:555:5
    frame #12: 0x00000001015539fc libcoreclr.dylib`SegmentScanByTypeChain(pSegment=0x00000001001b0000, uType=3, pfnBlockHandler=(libcoreclr.dylib`BlockScanBlocksWithoutUserData(TableSegment*, unsigned int, unsigned int, ScanCallbackInfo*) at handletablescan.cpp:541), pInfo=0x000000016ff9d858)(TableSegment*, unsigned int, unsigned int, ScanCallbackInfo*), ScanCallbackInfo*) at handletablescan.cpp:1595:13
    frame #13: 0x0000000101553c90 libcoreclr.dylib`TableScanHandles(pTable=0x0000000102810210, puType=0x000000016ff9d970, uTypeCount=1, pfnSegmentIterator=(libcoreclr.dylib`FullSegmentIterator(HandleTable*, TableSegment*, Holder<CrstBase*, &(CrstBase::AcquireLock(CrstBase*)), &(CrstBase::ReleaseLock(CrstBase*)), 0ul, &(int CompareDefault<CrstBase*>(CrstBase*, CrstBase*)), true>*) at handletablescan.cpp:1418), pfnBlockHandler=(libcoreclr.dylib`BlockScanBlocksWithoutUserData(TableSegment*, unsigned int, unsigned int, ScanCallbackInfo*) at handletablescan.cpp:541), pInfo=0x000000016ff9d858, pCrstHolder=0x000000016ff9d838)(HandleTable*, TableSegment*, Holder<CrstBase*, &(CrstBase::AcquireLock(CrstBase*)), &(CrstBase::ReleaseLock(CrstBase*)), 0ul, &(int CompareDefault<CrstBase*>(CrstBase*, CrstBase*)), true>*), void (*)(TableSegment*, unsigned int, unsigned int, ScanCallbackInfo*), ScanCallbackInfo*, Holder<CrstBase*, &(CrstBase::AcquireLock(CrstBase*)), &(CrstBase::ReleaseLock(CrstBase*)), 0ul, &(int CompareDefault<CrstBase*>(CrstBase*, CrstBase*)), true>*) at handletablescan.cpp:1712:17
    frame #14: 0x000000010154dcf0 libcoreclr.dylib`HndScanHandlesForGC(hTable=0x0000000102810210, scanProc=(libcoreclr.dylib`PinObject(Object**, unsigned long*, unsigned long, unsigned long) at objecthandle.cpp:256), param1=6173612744, param2=4318092408, types=0x000000016ff9d970, typeCount=1, condemned=2, maxgen=2, flags=0)(Object**, unsigned long*, unsigned long, unsigned long), unsigned long, unsigned long, unsigned int const*, unsigned int, unsigned int, unsigned int, unsigned int) at handletable.cpp:789:9
    frame #15: 0x000000010155779c libcoreclr.dylib`Ref_TracePinningRoots(condemned=2, maxgen=2, sc=0x000000016ff9dac8, fn=(libcoreclr.dylib`WKS::GCHeap::Promote(Object**, ScanContext*, unsigned int) at gc.cpp:36299))(Object**, ScanContext*, unsigned int)) at objecthandle.cpp:1027:21
    frame #16: 0x000000010155c254 libcoreclr.dylib`GCScan::GcScanHandles(fn=(libcoreclr.dylib`WKS::GCHeap::Promote(Object**, ScanContext*, unsigned int) at gc.cpp:36299), condemned=2, max_gen=2, sc=0x000000016ff9dac8)(Object**, ScanContext*, unsigned int), int, int, ScanContext*) at gcscan.cpp:168:9
    frame #17: 0x00000001015f9d38 libcoreclr.dylib`WKS::gc_heap::mark_phase(condemned_gen_number=2, mark_only_p=NO) at gc.cpp:20781:13
    frame #18: 0x00000001015f6220 libcoreclr.dylib`WKS::gc_heap::gc1() at gc.cpp:16694:13
    frame #19: 0x00000001016046a8 libcoreclr.dylib`WKS::gc_heap::garbage_collect(n=2) at gc.cpp:18279:9
    frame #20: 0x00000001015ebe48 libcoreclr.dylib`WKS::GCHeap::GarbageCollectGeneration(this=0x0000000100317800, gen=2, reason=reason_induced) at gc.cpp:37754:9
    frame #21: 0x0000000101630680 libcoreclr.dylib`WKS::GCHeap::GarbageCollectTry(this=0x0000000100317800, generation=2, low_memory_p=NO, mode=2) at gc.cpp:37012:12
    frame #22: 0x0000000101631474 libcoreclr.dylib`WKS::GCHeap::GarbageCollect(this=0x0000000100317800, generation=2, low_memory_p=false, mode=2) at gc.cpp:36946:30
    frame #23: 0x000000010150df78 libcoreclr.dylib`ETW::GCLog::ForceGCForDiagnostics() at eventtrace.cpp:969:44
    frame #24: 0x000000010150de20 libcoreclr.dylib`ETW::GCLog::ForceGC(l64ClientSequenceNumber=0) at eventtrace.cpp:875:5
    frame #25: 0x00000001015198fc libcoreclr.dylib`EtwCallbackCommon(ProviderIndex=DotNETRuntime, ControlCode=0, Level='\x05', MatchAnyKeyword=26738689, pFilterData=0x0000000000000000, isEventPipeCallback=YES) at eventtrace.cpp:4412:9
    frame #26: 0x00000001015199e0 libcoreclr.dylib`EventPipeEtwCallbackDotNETRuntime(SourceId=0x0000000000000000, ControlCode=0, Level='\x05', MatchAnyKeyword=26738689, MatchAllKeyword=0, FilterData=0x0000000000000000, CallbackContext=0x0000000000000000) at eventtrace.cpp:4449:5
    frame #27: 0x00000001013c5a24 libcoreclr.dylib`EventPipeProvider::InvokeCallback(eventPipeProviderCallbackData=<unavailable>) at eventpipeprovider.cpp:262:9
    frame #28: 0x00000001013b05ac libcoreclr.dylib`EventPipe::InvokeCallback(eventPipeProviderCallbackData=EventPipeProviderCallbackData @ 0x000000016ff9eb28) at eventpipe.cpp:1050:5
    frame #29: 0x00000001013aee64 libcoreclr.dylib`void EventPipe::RunWithCallbackPostponed<EventPipe::DisableHelper(unsigned long)::$_1>(f=(anonymous class) @ 0x000000016ff9ebb8)::$_1) at eventpipe.h:142:13
    frame #30: 0x00000001013adc90 libcoreclr.dylib`EventPipe::DisableHelper(id=4298257536) at eventpipe.cpp:484:5
    frame #31: 0x00000001013ae870 libcoreclr.dylib`EventPipe::Disable(id=4298257536) at eventpipe.cpp:470:5
    frame #32: 0x00000001013c3214 libcoreclr.dylib`EventPipeProtocolHelper::StopTracing(message=0x000000016ff9eee8, pStream=0x000000010032dcd0) at eventpipeprotocolhelper.cpp:201:5
  * frame #33: 0x00000001013c2b6c libcoreclr.dylib`EventPipeProtocolHelper::HandleIpcMessage(message=0x000000016ff9eee8, pStream=0x000000010032dcd0) at eventpipeprotocolhelper.cpp:130:9
    frame #34: 0x0000000101384e5c libcoreclr.dylib`DiagnosticServer::DiagnosticsServerThread((null)=0x0000000000000000) at diagnosticserver.cpp:83:17
    frame #35: 0x00000001010acdfc libcoreclr.dylib`CorUnix::CPalThread::ThreadEntry(pvParam=0x000000010280e000) at thread.cpp:1845:16
    frame #36: 0x00000001d528d0f8 libsystem_pthread.dylib`_pthread_start + 320
```

**Comment by sdmaclea** [comment]
GC/Scenarios/FinalizeTimeout/FinalizeTimeout/FinalizeTimeout.sh sometime hits a similar assert.  Unable to reproduce failure in lldb...
```
Assert failure(PID 52129 [0x0000cba1], Thread: 752942 [0xb7d2e]): RawGetMethodTable()
    File: /Users/steve/git/runtime/src/coreclr/src/gc/gc.cpp Line: 3699
    Image: /Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/Tests/Core_Root/corerun
```
Also saw this failure (which tends to be related to gc holes.)
```
Assert failure(PID 52113 [0x0000cb91], Thread: 750720 [0xb7480]): m_alignpad == 0
    File: /Users/steve/git/runtime/src/coreclr/src/vm/syncblk.cpp Line: 2966
    Image: /Users/steve/git/runtime/artifacts/tests/coreclr/OSX.arm64.Debug/Tests/Core_Root/corerun
```

**Comment by sdmaclea** [comment]
Fixed by #46201
